### ISSUE TITLE
Calico failing to deploy on a Workload cluster flavor=private

### ISSUE THREAD

**Issue by sshiba** [kubernetes-sigs/cluster-api-provider-azure#1607]
/kind bug

**What steps did you take and what happened:**
Deployed a workload cluster flavor=private following the steps in https://cluster-api.sigs.k8s.io/user/quick-start.html.

1. Deploy the management cluster: `kind create cluster`
2. Initialize management cluster with CAPI/CAPZ components: `clusterctl init --infrastructure azure`
3. Generate the deployment manifest
```
clusterctl generate cluster target-cluster \
  --kubernetes-version v1.21.1 \
  --control-plane-machine-count=1 \
  --worker-machine-count=3 \
  --flavor private \
  > deploy-private-target-cluster.yaml
```
4. Added the Secret for the Azure Client_Secret to `deploy-private-target-cluster.yaml`
5. From the management cluster environment, executed the attached `create-peering-vnet.txt`. Making sure info in the script matches `AzureCluster` resource defined in `deploy-private-target-cluster.yaml`.
6. Started the deployment with `kubectl apply -f ./deploy-private-target-cluster.yaml`
7. Once the cluster is deployed successfully, retrieve the workload cluster's kubeconfig with `clusterctl get kubeconfig private-target-cluster > private-target-cluster.kubeconfig`.
8. Deploy CNI with `kubectl --kubeconfig ./private-target-cluster.kubeconfig apply -f https://raw.githubusercontent.com/kubernetes-sigs/cluster-api-provider-azure/master/templates/addons/calico.yaml`

At this point calico-kube-controllers pod start crashing as shown below:
```
$ kubectl --kubeconfig private-target-cluster.kubeconfig get po -A
NAMESPACE     NAME                                                                 READY   STATUS             RESTARTS   AGE
kube-system   calico-kube-controllers-7b4dfb98f6-95jq7                             0/1     CrashLoopBackOff   4          2m50s
kube-system   calico-node-bj2hd                                                    1/1     Running            0          2m50s
kube-system   calico-node-jk5cs                                                    1/1     Running            0          2m50s
kube-system   calico-node-r4shs                                                    1/1     Running            0          2m50s
kube-system   calico-node-s98lv                                                    1/1     Running            0          2m50s
kube-system   coredns-558bd4d5db-gkssl                                             0/1     Running            0          14m
kube-system   coredns-558bd4d5db-xvw7n                                             0/1     Running            0          14m
kube-system   etcd-private-target-cluster-control-plane-6w5tc                      1/1     Running            0          14m
kube-system   kube-apiserver-private-target-cluster-control-plane-6w5tc            1/1     Running            0          14m
kube-system   kube-controller-manager-private-target-cluster-control-plane-6w5tc   1/1     Running            0          14m
kube-system   kube-proxy-6d558                                                     1/1     Running            0          13m
kube-system   kube-proxy-6rnsm                                                     1/1     Running            0          13m
kube-system   kube-proxy-cb26h                                                     1/1     Running            0          13m
kube-system   kube-proxy-sq2gx                                                     1/1     Running            0          14m
kube-system   kube-scheduler-private-target-cluster-control-plane-6w5tc            1/1     Running            0          14m
```

**What did you expect to happen:**
At this point I expect to have `calico-kube-controllers` pod be in Running state as well as the `coredns` pods. 

**Anything else you would like to add:**
Used this script to setup test environment: [create-peering-vnet.txt](https://github.com/kubernetes-sigs/cluster-api-provider-azure/files/6979058/create-peering-vnet.txt).
This script creates the VNet, subnets, nsg, route table, private DNS zones (record for apiserver internal FQDN), VNet peering and link to the private DNS.

**Environment:**
- cluster-api-provider-azure version: v0.5.1
- Kubernetes version: (use `kubectl version`): v1.21.1
- OS (e.g. from `/etc/os-release`): Ubuntu 20.04

**Comment by CecileRobertMichon** [comment]
trying to repro this 

/assign

**Comment by CecileRobertMichon** [comment]
@sshiba I could not repro:

this is a private cluster running the latest Calico in the main branch (https://raw.githubusercontent.com/kubernetes-sigs/cluster-api-provider-azure/main/templates/addons/calico.yaml):

```
capi@capz-e2e-0x0ujj-private-control-plane-cbmpg:~$ kubectl get nodes
NAME                                          STATUS   ROLES                  AGE     VERSION
capz-e2e-0x0ujj-private-control-plane-cbmpg   Ready    control-plane,master   7m20s   v1.21.2
capz-e2e-0x0ujj-private-control-plane-jjvd5   Ready    control-plane,master   10m     v1.21.2
capz-e2e-0x0ujj-private-control-plane-wrshw   Ready    control-plane,master   9m18s   v1.21.2
capz-e2e-0x0ujj-private-md-0-fzt64            Ready    <none>                 9m18s   v1.21.2

capi@capz-e2e-0x0ujj-private-control-plane-cbmpg:~$ kubectl get pods -A
NAMESPACE     NAME                                                                  READY   STATUS    RESTARTS   AGE
kube-system   calico-kube-controllers-846b5f484d-6xk8q                              1/1     Running   0          10m
kube-system   calico-node-dnsf4                                                     1/1     Running   0          7m34s
kube-system   calico-node-lf5t7                                                     1/1     Running   0          9m32s
kube-system   calico-node-n4gqd                                                     1/1     Running   0          10m
kube-system   calico-node-qwmd7                                                     1/1     Running   0          9m32s
kube-system   coredns-558bd4d5db-649mw                                              1/1     Running   0          10m
kube-system   coredns-558bd4d5db-zvd64                                              1/1     Running   0          10m
kube-system   etcd-capz-e2e-0x0ujj-private-control-plane-cbmpg                      1/1     Running   0          7m32s
kube-system   etcd-capz-e2e-0x0ujj-private-control-plane-jjvd5                      1/1     Running   0          10m
kube-system   etcd-capz-e2e-0x0ujj-private-control-plane-wrshw                      1/1     Running   0          9m25s
kube-system   kube-apiserver-capz-e2e-0x0ujj-private-control-plane-cbmpg            1/1     Running   0          7m33s
kube-system   kube-apiserver-capz-e2e-0x0ujj-private-control-plane-jjvd5            1/1     Running   0          10m
kube-system   kube-apiserver-capz-e2e-0x0ujj-private-control-plane-wrshw            1/1     Running   0          9m25s
kube-system   kube-controller-manager-capz-e2e-0x0ujj-private-control-plane-cbmpg   1/1     Running   0          7m33s
kube-system   kube-controller-manager-capz-e2e-0x0ujj-private-control-plane-jjvd5   1/1     Running   1          10m
kube-system   kube-controller-manager-capz-e2e-0x0ujj-private-control-plane-wrshw   1/1     Running   0          9m25s
kube-system   kube-proxy-bgdm2                                                      1/1     Running   0          9m32s
kube-system   kube-proxy-d4mr7                                                      1/1     Running   0          7m34s
kube-system   kube-proxy-hfd77                                                      1/1     Running   0          10m
kube-system   kube-proxy-lwsrj                                                      1/1     Running   0          9m32s
kube-system   kube-scheduler-capz-e2e-0x0ujj-private-control-plane-cbmpg            1/1     Running   0          7m33s
kube-system   kube-scheduler-capz-e2e-0x0ujj-private-control-plane-jjvd5            1/1     Running   1          10m
kube-system   kube-scheduler-capz-e2e-0x0ujj-private-control-plane-wrshw            1/1     Running   0          9m25s
```

I wonder if something is wrong with kubelet on your cluster? Is kubelet in running state? Otherwise could it be a DNS issue because of the vnet peering?

**Comment by sshiba** [comment]
@CecileRobertMichon, I tried with 3x control plane nodes and 1x worker node and after a while calico-kube-controller get into Ready state. The coredns seems not very healthy:

**Comment by sshiba** [comment]
@CecileRobertMichon, would it be possible to share the script that you used for setting up the workload cluster's vnet and vnet peering?

**Comment by CecileRobertMichon** [comment]
@sshiba apologies for the delayed response. Here are the steps I followed (very similar to yours):

1) create an AKS cluster with --network-plugin "azure" (kubenet doesn't work with AAD pod identity) to use as management cluster
2) export variables required by your script. Ephemeral RG and vnet are the AKS rg and vnet from above.
3) run your script (without any changes to it)
4) update the template you pasted above to add my vnet and subnet CIDR ranges and the internal LB IP 
5) on the AKS cluster, run `clusterctl init --infrastructure azure`, then generate the template with the command you had above and apply it with `kubectl apply -f ./deploy-private-target-cluster.yaml`
6) from a VM in the cluster vnet, I got the kubeconfig and ran `kubectl --kubeconfig ./private-target-cluster.kubeconfig apply -f https://raw.githubusercontent.com/kubernetes-sigs/cluster-api-provider-azure/master/templates/addons/calico.yaml`

The result is a healthy private cluster with Calico running.
